EIPS/eip-665.md

```diff
@@ -0,0 +1,82 @@
+---
+eip: 665
+title: Add precompiled contract for Ed25519 signature verification
+author: Tobias Oberstein <tobias.oberstein@crossbario.com>
+status: Draft
+type: Standards Track
+category: Core
+created: 2018-03-25
+---
+
+## Simple Summary
+
+Support performant and cheap verification of Ed25519 cryptographic signatures in smart contracts in general by adding a precompiled contract for Ed25519 signature verification to the EVM.
+
+## Abstract
+
+Verification of Ed25519 cryptographic signatures is obviously possible in EVM bytecode. However, the gas cost will be very high, and computationally expensive, as such tight, wide word operations intensive code as required for Ed25519 is not a good fit for the EVM bytecode model.
+
+The addition of a native compiled function, in a precompiled contract, to the EVM solves both cost and performance problems.
+
+## Motivation
+
+One motivation for Ed25519 signature verification in smart contracts is to associate existing off-chain systems, records or accounts that use Ed25519 with blockchain transactions.
+
+Another motivation is the processing of external, Ed25519 proof-of-stake based blockchains within Ethereum smart contracts.
+
+When a transactions contains data that comes with an Ed25519 signature, that proves that the sender of the Ethereum transaction was also in control of the private key (and the data), and this allows the contract to establish an association between the blockchain and the external system or account, and the external system establish the reverse relation.
+
+For example, a contract might check a Ed25519 signed piece of data submitted to the Ethereum transaction like the current block number. That proves to the contract, that the sender is in possession of both the Ethereum private key and the Ed25518 private key, and hence the contract will accept an association between both. This again can be the root anchor for various powerful applications, as now a potentially crypto holding key owner has proven to be in control of some external off-chain system or account, like e.g. a DNS server, a DNS domain, a cluster node and so on.
+
+## Specification
+
+The proposal adds a new precompiled function with the following signature
+
+```
+ed25519verify(bytes32 m, bytes32 pk, bytes32 s1, bytes32 s2) returns (uint8)
```

Comment:
Existing precompiles do not use function selectors (for example, the RSA precompile). I think it'd be a good idea to remain consistent with that, and specify this in terms of call data.

Reply:
Not sure what you mean, could you expand? Note that `m` is supposed to be the message to be signed, and is needed by the actual function body.

Reply:
Take a look at http://eips.ethereum.org/EIPS/eip-196 for instance - the precompile takes call data that is the direct encoding of the parameters, with no function selector (eg, it's not a Solidity-standard multi-function contract). All precompiles so far are like this; it'd be a good idea to retain consistency here.

Reply:
Ok, got it. Thanks for pointing me in the right direction! So I believe the changes I pushed should address that.